Tests/DatadogTests/Datadog/Tracing/Utils/URLSessionSwizzlerTests.swift

```diff
@@ -7,107 +7,242 @@
+            timeout: 0.1,
+            enforceOrder: true
+        )
     }
 
-    func test_unswizzleDataTaskWithURL() {
-        let session = URLSession(configuration: .default)
+    func test_dataTask_requestCompletion_alwaysIntercept() {
+        let completionExpectation = XCTestExpectation(description: "completionExpectation")
+        let interceptorTuple = buildInterceptorTuple(modifiedURL: modifiedURL)
 
-        XCTAssertNoThrow(try URLSessionSwizzler.swizzleDataTaskWithURL())
+        XCTAssertNoThrow(try swizzleURLSession(interceptor: interceptorTuple.block))
 
-        let url = URL(string: "http://foo.bar")!
-        let task = session.dataTask(with: url)
+        let task = session.dataTask(with: URLRequest.mockAny()) { _, _, _ in
+            completionExpectation.fulfill()
+        }
 
-        XCTAssertEqual(task.originalRequest?.url, url)
+        XCTAssertNotEqual(task.originalRequest?.url, URL.mockAny())
+        XCTAssertEqual(task.originalRequest?.url, modifiedURL)
 
-        try! MethodSwizzler.shared.unswizzle(
-            selector: URLSessionSwizzler.Selectors.DataTaskWithURL,
-            in: URLSessionSwizzler.subjectClass
+        wait(
+            for: [interceptorTuple.interceptionExpectation],
+            timeout: 0.1,
+            enforceOrder: true
         )
 
-        let unswizzledTask = session.dataTask(with: url)
-        XCTAssertEqual(unswizzledTask.originalRequest?.url, url)
+        task.resume()
+
+        let resumeExpectations: [XCTestExpectation] = [
+            interceptorTuple.observationStartingExpectation,
+            completionExpectation,
+            interceptorTuple.observationCompletedExpectation
+        ]
+        wait(
+            for: resumeExpectations,
+            timeout: 0.1,
+            enforceOrder: true
+        )
     }
 
-    func test_swizzleDataTaskWithRequest() {
-        defer {
-            try! MethodSwizzler.shared.unswizzle(
-                selector: URLSessionSwizzler.Selectors.DataTaskWithRequest,
-                in: URLSessionSwizzler.subjectClass
-            )
+    func test_dataTask_requestCompletion_alwaysIntercept_noResume() {
+        let modifiedURL = URL(string: "https://foo.bar.modified")!
+        let interceptor: RequestInterceptor = { originalRequest in
+            var modifiedReq = originalRequest
+            modifiedReq.url = modifiedURL
+
+            let observer: TaskObserver = { event in
+                XCTFail("Observer should not be called without resume()")
+            }
+            return (request: modifiedReq, taskPayload: observer)
         }
 
-        let session = URLSession(configuration: .default)
+        XCTAssertNoThrow(try swizzleURLSession(interceptor: interceptor))
+
+        let task = session.dataTask(with: URLRequest.mockAny()) { _, _, _ in }
+
+        XCTAssertEqual(task.originalRequest?.url, modifiedURL)
+    }
+}
+
+// MARK: - 3rd party swizzlers
+
+class ThirdPartySwizzlingTests: XCTestCase {
```

Comment:
If we need to test 3rd party swizzling against a specific swizzler implementation: `URLSessionSwizzler`, does it mean we'll need it for other swizzlerts as well?
* `UIViewControllerSwizzler` + `ThirdPartyUIViewControllerSwizzlingTests`,
* `UIViewNavigationControllerSwizzler` + `ThirdPartyUIViewNavigationControllerSwizzlerTests`,

Reply:
correct, 3rd party swizzling tests make sure that new implementation doesn't swallow old implementation of method. wherever we provide a new implementation for a method, ideally we should test it with a 3rd party swizzler

but when needed, we can make 3rd party swizzler more generic
currently it's more of a simplicity-first class

Reply:
👍 ok, let's keep it concrete type now and we'll try to make it generic in future use cases.

Reply:
i'm assuming this discussion is resolved by now ✅